### PR TITLE
Micro API Gateway related components

### PR DESCRIPTION
**Pull request by chalitha1989**
### Proposed changes in this pull request

This PR introduces following components:

- A throttling synchronizing component for syncing throttling policy changes with the micro gateway.
- A usage publishing component for uploading API usage data from the micro gateway server to the publisher node and publisher not to analytics server.
- A web app for facilitating usage data upload and API lifecycly update retrieval through REST APIs.

### When should this PR be merged

This PR has no external dependencies

### Follow up actions

NA

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/test/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPITest.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        PowerMockito.spy(UploadedUsageFileInfoDAO.class);
+        PowerMockito.doNothing()
+                .when(UploadedUsageFileInfoDAO.class, "persistFileUpload", any(UploadedFileInfoDTO.class),
+                        any(InputStream.class));
+        MicroGatewayRestAPI microGwAPI = new MicroGatewayRestAPI();
+        Response response = microGwAPI.uploadFile(anyInputStream, httpHeaders);
+        Assert.assertEquals(201, response.getStatus());
+    }
+
+        @Test
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/test/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPITest.java]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.usage.publisher/pom.xml [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <properties>
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.usage.publisher/pom.xml]: Shall we move the versions to root pom?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.usage.publisher/pom.xml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        <orbit.httpmime.version>4.3.1.wso2v2</orbit.httpmime.version>
+        <xml.apis.version>1.4.01</xml.apis.version>
+        <com.lmax.disruptor.version.range>[2.10.0,3.4)</com.lmax.disruptor.version.range>
+        <lmax.disruptor.version.range>[2.10.0,3.4)</lmax.disruptor.version.range>
+        <databridge.agent.range>[5.1,6)</databridge.agent.range>
+        <slf4j.wso2.version>1.5.10.wso2v1</slf4j.wso2.version>
+    </properties>
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.usage.publisher/pom.xml]: Shall we move the versions to root pom?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ConditionalGroupDTO.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2017 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ConditionalGroupDTO.java]: Shall we make it 2018? There are several other places we need to fix.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/pom.xml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-war-plugin</artifactId>
+                <version>2.2</version>
+                <configuration>
+                    <packagingExcludes>WEB-INF/lib/*.jar</packagingExcludes>
+                    <warName>micro-gateway</warName>
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/pom.xml]: Better to include a version too. eg: micro-gateway#0.9

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPI.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                if (log.isDebugEnabled()) {
+                    log.debug("Successfully retrieved identifiers of updated APIs for Tenant : "
+                            + tenantDomain + " By : " + tenantAwareUsername);
+                }
+                return Response.ok().entity(obj.toString()).build();
+            } else {
+                log.warn("Unauthorized access for API publish/re-publish event publishing service. "
+                        + authDTO.getMessage());
+                JsonObject responseObj = new JsonObject();
+                responseObj.addProperty("message", authDTO.getMessage());
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPI.java]: Shall we move "message" to a Constant?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPI.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                responseObj.addProperty("description", authDTO.getDescription());
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/MicroGatewayRestAPI.java]: Same here for "description"

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/test/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/TestUtil.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        this.setCarbonHome();
+    }
+
+    private void setCarbonHome() throws Exception {
+        if (StringUtils.isEmpty(System.getProperty("carbon.home"))) {
+            String filePath =
+                    "src" + File.separator + "test" + File.separator + "resources" + File.separator + "carbon-home";
+            File file = new File(filePath);
+            if (file.exists()) {
+                System.setProperty("carbon.home", file.getAbsolutePath());
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/test/java/org/wso2/carbon/apimgt/micro/gateway/rest/api/TestUtil.java]: Shall we use a constant here for "carbon.home"?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+@ApiModel(description = "")
+public class AdvancedThrottlePolicyDTO extends ThrottlePolicyDTO {
+  
+  
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java]: Let's remove the additional lines in this class

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  @JsonProperty("conditionalGroups")
+  public List<ConditionalGroupDTO> getConditionalGroups() {
+    return conditionalGroups;
+  }
+  public void setConditionalGroups(List<ConditionalGroupDTO> conditionalGroups) {
+    this.conditionalGroups = conditionalGroups;
+  }
+
+  
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java]: Same here

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ApplicationThrottlePolicyListDTO.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  @ApiModelProperty(value = "")
+  @JsonProperty("list")
+  public List<ApplicationThrottlePolicyDTO> getList() {
+    return list;
+  }
+  public void setList(List<ApplicationThrottlePolicyDTO> list) {
+    this.list = list;
+  }
+
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ApplicationThrottlePolicyListDTO.java]: Let's remove the additional lines here as well. Please do the same for the rest of the classes.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/test/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/TestUtil.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        this.setCarbonHome();
+    }
+
+    private void setCarbonHome() throws Exception {
+        if (StringUtils.isEmpty(System.getProperty("carbon.home"))) {
+            String filePath =
+                    "src" + File.separator + "test" + File.separator + "resources" + File.separator + "carbon-home";
+            File file = new File(filePath);
+            if (file.exists()) {
+                System.setProperty("carbon.home", file.getAbsolutePath());
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/test/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/TestUtil.java]: Use constatnts

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ConditionalGroupDTO.java (hunk 4 above)
Comment (chalitha1989) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/ConditionalGroupDTO.java]: These dto classes were taken from the carbon apimgt and therefore we didn't do any modifications to them. I will correct the year in the license header.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java (hunk 9 above)
Comment (chalitha1989) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.throttling.synchronizer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/throttling/synchronizer/dto/AdvancedThrottlePolicyDTO.java]: DTO classes were taken from the apimgt repo and we didn't change those. Will do the requested changes.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/webapp/META-INF/webapp-classloading.xml [hunk 13]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+
+<!--
+  ~ Copyright 2015 WSO2, Inc. (http://wso2.com) All Rights Reserved.
```
Comment (NipunaPrashan) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.rest.api/src/main/webapp/META-INF/webapp-classloading.xml]: Let's change the year accordingly. This should be 2018

**Pull request merged**